### PR TITLE
nix-repl: use history file

### PR DESCRIPTION
**Pull request by leungbk**
----

#

### CODE REVIEW

File: nix-repl.el [hunk 1]
```diff
@@ -34,10 +34,30 @@
...
+    (with-current-buffer buf
+      (when (eq major-mode 'nix-repl-mode)
+        (comint-write-input-ring)))))
+
 (define-derived-mode nix-repl-mode comint-mode "Nix-REPL"
   "Interactive prompt for Nix."
   (setq-local comint-prompt-regexp nix-prompt-regexp)
   (setq-local comint-prompt-read-only t)
+  (let* ((is-remote (file-remote-p default-directory))
+         (maybe-xdg-data-home (shell-command-to-string "echo -n $XDG_DATA_HOME"))
```
Comment (matthewbauer) [inline nix-repl.el:49]: use `getenv` to avoid a subprocess

File: nix-repl.el [hunk 2]
```diff
@@ -34,10 +34,30 @@
...
+         (path-prefix (if (string-empty-p maybe-xdg-data-home)
+                          "~/.local/share"
+                        maybe-xdg-data-home))
+         (history-path (concat
+                        is-remote
+                        path-prefix
+                        "/nix/repl-history")))
+    (setq-local comint-input-ring-file-name history-path))
+  (comint-read-input-ring t)
+  (add-hook 'kill-buffer-hook #'comint-write-input-ring nil 'local)
```
Comment (matthewbauer) [inline nix-repl.el:59]: shell-mode does this like this:

```
(set-process-sentinel (get-buffer-process (current-buffer))
				#'shell-write-history-on-exit)
```

Do you think that might be a better hook? (replacing shell-write-history-on-exit with comint-write-input-ring that is) That would write out after `:q` or `Ctrl-D`.

File: nix-repl.el (hunk 1 above)
Comment (leungbk) [reply nix-repl.el]: OK, I used `getenv` for non-remote. I don't really know how things work in remote settings; I saw that Emacs' `shell.el` used a subprocess for remote settings and `getenv` for local settings:

``` elisp
(when remote
        ;; `shell-snarf-envar' does not work trustworthy.
        (setq hsize (shell-command-to-string "echo -n $HISTSIZE")
              hfile (shell-command-to-string "echo -n $HISTFILE")))
```

So I updated this patch to do the same thing. Is that the right way of doing this?

File: nix-repl.el (hunk 2 above)
Comment (leungbk) [reply nix-repl.el:62]: With what I submitted (which I took from [`racket-mode`'s REPL](https://github.com/greghendershott/racket-mode/blob/337d8f60c90a69b12c64c579c628f2388e16a7cb/racket-repl.el#L996)), I noticed that it was already writing out after `:q` and `Ctrl-D`. But I couldn't figure out how to write out after `C-x k` in the Nix REPL without modifying `kill-buffer-hook`, and using `(set-process-sentinel (get-buffer-process (current-buffer)) #'comint-write-input-ring)` in place of this does not appear to write after `C-x k`.

**Pull request merged**